### PR TITLE
feat: add remote invoke factories

### PR DESCRIPTION
**Pull request by mndeveci**
#### Which issue(s) does this change fix?
https://github.com/aws/aws-sam-cli/discussions/5018

This PR adds some library implementations for Remote Invoke feature which contains;
- `RemoteInvokeExecutorFactory` implementation which will then be used to instantiate `RemoteInvokeExecutor` instance depending on the resource type. It is mapping is empty now, as implementations will be added later.
- `RemoteInvokeExecutionInfo` object which holds request and response information as well as any exception that is been raised during the execution.
- Abstract `RemoteInvokeRequestResponseMapper` definition, which then be used to implement different mappers for different remote invoke use case.
  - `ResponseObjectToJsonStringMapper` implements mapping an object into JSON string.
- Abstract `BotoActionExecutor` definition, which then be extended by different remote invoke instances, by implementing `_execute_action` and `_execute_action_file` methods.
- `RemoteInvokeExecutor` definition which is a container of holding request mappers, response mappers and `BotoActionExecutor` instance, which can be used from remote invoke context later.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/remote_invoke/remote_invoke_executors.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    Or for a response object, if it contains streaming results, a mapper can convert them back
+    to string to display on
+    """
+
+    @abstractmethod
+    def map(self, remote_invoke_input: RemoteInvokeExecutionInfo) -> RemoteInvokeExecutionInfo:
+        raise NotImplementedError()
+
+
+class ResponseObjectToJsonStringMapper(RemoteInvokeRequestResponseMapper):
```
Comment (hnnasit) [inline samcli/lib/remote_invoke/remote_invoke_executors.py:66]: Can we add docstrings for the class and function here?

File: samcli/lib/remote_invoke/remote_invoke_executors.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        self,
+        request_mappers: List[RemoteInvokeRequestResponseMapper],
+        response_mappers: List[RemoteInvokeRequestResponseMapper],
+        boto_action_executor: BotoActionExecutor,
+    ):
+        self._request_mappers = request_mappers
+        self._response_mappers = response_mappers
+        self._boto_action_executor = boto_action_executor
+
+    def execute(self, remote_invoke_input: RemoteInvokeExecutionInfo) -> RemoteInvokeExecutionInfo:
```
Comment (hnnasit) [inline samcli/lib/remote_invoke/remote_invoke_executors.py:173]: docstring for this function.

File: tests/unit/lib/remote_invoke/test_remote_invoke_executors.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+from samcli.lib.remote_invoke.remote_invoke_executors import (
+    RemoteInvokeExecutionInfo,
+    BotoActionExecutor,
+    RemoteInvokeExecutor,
+    ResponseObjectToJsonStringMapper,
+    RemoteInvokeRequestResponseMapper,
+)
+
+
+class TestTestExecutionInfo(TestCase):
```
Comment (hnnasit) [inline tests/unit/lib/remote_invoke/test_remote_invoke_executors.py:16]:
```suggestion
class TestRemoteInvokeExecutionInfo(TestCase):
```

File: tests/unit/lib/remote_invoke/test_remote_invoke_executors.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            patched_execute_action.side_effect = given_exception
+
+            result = self.boto_action_executor.execute(test_execution_info)
+
+            patched_execute_action.assert_called_with(given_payload)
+
+            self.assertEqual(given_exception, result.exception)
+
+
+class TestTestExecutor(TestCase):
```
Comment (hnnasit) [inline tests/unit/lib/remote_invoke/test_remote_invoke_executors.py:115]:
```suggestion
class TestRemoteInvokeExecutor(TestCase):
```

File: tests/unit/lib/remote_invoke/test_remote_invoke_executor_factory.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+from unittest import TestCase
+from unittest.mock import patch, Mock
+
+from samcli.lib.remote_invoke.remote_invoke_executor_factory import RemoteInvokeExecutorFactory
+
+
+class TestTestExecutorFactory(TestCase):
```
Comment (hnnasit) [inline tests/unit/lib/remote_invoke/test_remote_invoke_executor_factory.py:7]:
```suggestion
class TestRemoteInvokeExecutorFactory(TestCase):
```

**Review by hnnasit — CHANGES_REQUESTED** [review]

**Review by hnnasit — APPROVED** [review]
LGTM! Need to run `make black`.

File: samcli/lib/remote_invoke/remote_invoke_executors.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    exception: Optional[Exception]
+
+    def __init__(self, payload: Optional[Union[str, List, dict]], payload_file: Optional[TextIOWrapper]):
+        self.payload = payload
+        self.payload_file = payload_file
+        self.response = None
+        self.exception = None
+
+    def is_file_provided(self):
+        return self.payload_file
```
Comment (moelasmar) [inline samcli/lib/remote_invoke/remote_invoke_executors.py:40]: what is the return type of this method? .. if it is boolean, could you please change the return statement to be `bool(elf.payload_file)`

Comment (moelasmar) [reply samcli/lib/remote_invoke/remote_invoke_executors.py:40]: Also add the return type to the function definition

File: samcli/lib/remote_invoke/remote_invoke_executors.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+    @property
+    def payload_file_path(self) -> Optional[TextIOWrapper]:
+        return self.payload_file if self.is_file_provided() else None
+
+    def is_succeeded(self):
+        return self.response
```
Comment (moelasmar) [inline samcli/lib/remote_invoke/remote_invoke_executors.py:47]: could you please add the return type to the function definition. 
Is it possible to have a failure response?
can you modify the return statement to return boolean if the return type is boolean

**Review by moelasmar — APPROVED** [review]
LGTM, just left some minor comments

**Review by moelasmar — APPROVED** [review]
LGTM, just left some minor comments

File: samcli/lib/remote_invoke/remote_invoke_executors.py (hunk 7 above)
Comment (mndeveci) [reply samcli/lib/remote_invoke/remote_invoke_executors.py:47]: > Is it possible to have a failure response?
It is possible, and it is kept inside another field called `exception`. That field is only used in tests for now, will be used once we raised follow-up PRs.

**Pull request merged**